builtin/credential/approle/path_tidy_user_id.go

```diff
@@ -39,206 +40,224 @@ func (b *backend) tidySecretID(ctx context.Context, req *logical.Request) (*logi
+			// ExpirationTime not being set indicates non-expiring SecretIDs
+			if !result.ExpirationTime.IsZero() && time.Now().After(result.ExpirationTime) {
+				logger.Trace("found expired secret ID")
+				// Clean up the accessor of the secret ID first
+				err = b.deleteSecretIDAccessorEntry(ctx, s, result.SecretIDAccessor, secretIDPrefixToUse)
 				if err != nil {
-					return errwrap.Wrapf("failed to read secret ID accessor entry: {{err}}", err)
+					return errwrap.Wrapf("failed to delete secret ID accessor entry: {{err}}", err)
 				}
-				if accessorEntry == nil {
-					logger.Trace("found nil accessor")
-					if err := s.Delete(ctx, entryIndex); err != nil {
-						return errwrap.Wrapf(fmt.Sprintf("error deleting secret ID %q from storage: {{err}}", secretIDHMAC), err)
-					}
-					return nil
+
+				if err := s.Delete(ctx, entryIndex); err != nil {
+					return errwrap.Wrapf(fmt.Sprintf("error deleting SecretID %q from storage: {{err}}", secretIDHMAC), err)
 				}
 
-				// ExpirationTime not being set indicates non-expiring SecretIDs
-				if !result.ExpirationTime.IsZero() && time.Now().After(result.ExpirationTime) {
-					logger.Trace("found expired secret ID")
-					// Clean up the accessor of the secret ID first
-					err = b.deleteSecretIDAccessorEntry(ctx, s, result.SecretIDAccessor, secretIDPrefixToUse)
-					if err != nil {
-						return errwrap.Wrapf("failed to delete secret ID accessor entry: {{err}}", err)
-					}
+				return nil
+			}
 
-					if err := s.Delete(ctx, entryIndex); err != nil {
-						return errwrap.Wrapf(fmt.Sprintf("error deleting SecretID %q from storage: {{err}}", secretIDHMAC), err)
-					}
+			// At this point, the secret ID is not expired and is valid. Flag
+			// the corresponding accessor as not needing attention.
+			skipHashes[salt.SaltID(result.SecretIDAccessor)] = true
 
-					return nil
-				}
+			return nil
+		}
 
-				// At this point, the secret ID is not expired and is valid. Delete
-				// the corresponding accessor from the accessorMap. This will leave
-				// only the dangling accessors in the map which can then be cleaned
-				// up later.
-				salt, err := b.Salt(ctx)
+		for _, roleNameHMAC := range roleNameHMACs {
+			logger.Trace("listing secret ID HMACs", "role_hmac", roleNameHMAC)
+			secretIDHMACs, err := s.List(ctx, fmt.Sprintf("%s%s", secretIDPrefixToUse, roleNameHMAC))
+			if err != nil {
+				return err
+			}
+			for _, secretIDHMAC := range secretIDHMACs {
+				err = secretIDCleanupFunc(secretIDHMAC, roleNameHMAC, secretIDPrefixToUse)
 				if err != nil {
 					return err
 				}
-				delete(accessorMap, salt.SaltID(result.SecretIDAccessor))
-
-				return nil
 			}
+		}
 
+		// Accessor indexes were not getting cleaned up until 0.9.3. This is a fix
+		// to clean up the dangling accessor entries.
+		if len(accessorHashes) > len(skipHashes) {
+			// There is some raciness here because we're querying secretids for
```

Comment:
I think this is fine because, the newer secret IDs that are created after the lock acquisition should ideally not have this issue since it does the cleanups properly. It is likely that the cleanups are applicable for very old secret IDs. I wouldn't be worried about not locking here.

Reply:
Yeah, I agree, but I thought it was worth calling out to the reader that it's known and deemed acceptable.  Thanks for validating.